### PR TITLE
Added bash script to automate signing .json files

### PR DESCRIPTION
**Pull request by jmjaffe37**

### CODE REVIEW

File: scripts/sign.sh
Comment (d3r3kk) [inline scripts/sign.sh:20]: Do we want to _check for all required parameters_ here and send the HELP if we are missing anything? Or, alternatively, would we want to set some reasonable defaults for the `content_dir`, `secret_key_file`, and `public_key_file` and make some assumptions?

File: scripts/sign.sh
Comment (d3r3kk) [inline scripts/sign.sh]: Might be useful to add meaningful defaults and communicate them here, or add a bit of text denoting if a param is required or not. (This is further to my other comment).

File: scripts/sign.sh
Comment (d3r3kk) [inline scripts/sign.sh:38]: That's a good tip, maybe add a quick sentence about what the loop does?

File: scripts/README.md
Comment (d3r3kk) [inline scripts/README.md:1]: Since this is a README that pertains only to the `sign.sh` file (but is in the generic `scripts/` folder) perhaps rename it to `scripts/sign.sh.README.md` or maybe make this a _section_ in the file (and make it clear where the next author of the next script might add their documentation?).

**Review by d3r3kk — CHANGES_REQUESTED** [review]
Looks pretty good @jmjaffe37 but I think we can improve a little bit on the documentation/clarity on useability a bit.

File: scripts/sign.sh
Comment (karianna) [inline scripts/sign.sh:4]: Are we missing a set -eux or similar?

Comment (d3r3kk) [reply scripts/sign.sh:4]: obligitory "gah, `bash`" comment.

Comment (d3r3kk) [reply scripts/sign.sh:4]: (@karianna's correct here @jmjaffe37)

File: scripts/sign.sh
Comment (gdams) [inline scripts/sign.sh]:
```suggestion
        -s|--secret-key) secret_key_file="$2"; shift ;;
```

File: scripts/sign.sh
Comment (gdams) [inline scripts/sign.sh]:
```suggestion
        -p|--public-key) public_key_file="$2"; shift ;;
```

File: scripts/sign.sh
Comment (gdams) [inline scripts/sign.sh]:
```suggestion
    openssl dgst -sha256 -sign "$secret_key_file" -out ${json}.sig $json
```

File: scripts/sign.sh
Comment (gdams) [inline scripts/sign.sh]:
```suggestion
    openssl dgst -sha256 -verify "$public_key_file" -signature ${json}.sig $json
```

**Review by gdams — CHANGES_REQUESTED** [review]
a few variables need quoting to prevent globbing

File: scripts/README.md
Comment (jmjaffe37) [reply scripts/README.md:1]: I think the second option was my intention. "Script usage" is meant to be the title of the README (meaning that it is the readme for all scripts in the script directory). The subheading "sign.sh" (on line 3) is meant to be the section that describes the usage for just the sign.sh file, and subsequent sections can be created below or above. Do you have suggestions on how I could improve this format for clarity?

File: scripts/sign.sh
Comment (jmjaffe37) [reply scripts/sign.sh:20]: I added defaults for `secret_key_file`, and `public_key_file`

File: scripts/sign.sh
Comment (jmjaffe37) [reply scripts/sign.sh:4]: I added this, but quick question:
Are we sure that we want to include the 'x' flag? It seems to make printed output weird, Especially when we get output from `--help`

File: scripts/sign.sh
Comment (jmjaffe37) [reply scripts/sign.sh:38]: I added more description within the loop. Let me know if this works :)

File: scripts/README.md
Comment (d3r3kk) [reply scripts/README.md:1]: No, I understand what you've done now on reading it a second time.

File: scripts/sign.sh
Comment (d3r3kk) [reply scripts/sign.sh:4]: Ah yes, `-x` is `xtrace` and I believe it's intended use is for _debugging_ so you might want to leave that one off. (You can turn it on temporarily from the command line by calling your script with `bash -x sign.sh`.)

But I'll defer to the `bash` experts in the crowd.

**Review by d3r3kk — APPROVED** [review]
LGTM, maybe get  sign-off from @gdams or @karianna before you merge.

File: scripts/sign.sh
Comment (jmjaffe37) [reply scripts/sign.sh:4]: @karianna @gdams

File: scripts/README.md
Comment (dsgrieve) [reply scripts/README.md:1]: Just my $.02 USD. I recommend putting each script into its own markdown file. In README.md, I'd have an abstract for the script and a link to script's markdown. This would make it easier to maintain, in my opinion, since it separates the doc for the specific script from what may grow to be a very large README.

File: scripts/sign.sh
Comment (dsgrieve) [inline scripts/sign.sh:31]: If there is a default, is the arg really required? And if it is the default, shouldn't the parameter reference below be ${secret_key_file:-../private.pem} ? Same for --public-key

Comment (jmjaffe37) [reply scripts/sign.sh:31]: I meant that it is required because the script will error if there is no key at the specified location. Microsoft has done this in other documentation, example: [DownloadPipelineArtifact task](https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/download-pipeline-artifact-v2?view=azure-pipelines)[.](https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/download-pipeline-artifact-v2?view=azure-pipelines).) Here, we can see that `buildType` is required, but has a default

Also, I am unsure what your comment is suggesting in the 'parameter reference' section

File: scripts/sign.sh
Comment (karianna) [reply scripts/sign.sh:4]: set -eu is fine

**Review by karianna — APPROVED** [review]

File: scripts/sign.sh
Comment (dsgrieve) [reply scripts/sign.sh:31]: I think `buildType` being required and having a default is confusing and I have submitted feedback on that page stating such. 

No where in the shell script is `../private.pem` used (other than being echoed in the usage). If I ran this script without  `--secret-key`, I would get an unbound variable error. But your usage says it has a default, so how can it be unbound? If I don't give `--secret-key`, shouldn't the script use `../private.pem`? That's what the usage says will happen. So, in my view, this is broken. If `--secret-key` is not given on the command line, then `secret_key_file` should default to `../private.pem`. The bash way to do that is with `${property:-default}` or `${property:=default}`. There is a subtle difference in the two which I encourage you to find out about in the bash man page. Look for details on bash parameter expansion.  

Of course, if your intent is to error with an `unbound variable` if `--secret-key` is missing, then remove the mention of the default (in both places, obviously). 

I'll toss this out there, too. Try this from a bash command line: `${secret_key_file:?--secret-key parameter is required}` and see what happens. Is that feedback more useful than "unbound variable"?

Lastly, there are ways to put error hooks into a bash script. Typically, these are used to show the command line usage. I'm not saying you need to do that here, but it is fun stuff to play with.

Comment (jmjaffe37) [reply scripts/sign.sh:31]: Thanks for the feedback David! I have a quick question about the error you are getting though: I am unable to reproduce the unbound variable error you mentioned. Here is the result of running the script both with and without the `--secret-key` variable explicitly defined. In both cases, the script uses `../private.pem` as expected without error (and I believe it would error if there was an unbound variable exception since I have `set -eu` at the top of the script):
image

Comment (dsgrieve) [reply scripts/sign.sh:31]: OMG. My bad! I totally didn't see 
```
# Set defaults
secret_key_file="../private.pem"
public_key_file="../public.pem"
```
I still stand by my point about "required" and having a default.

**Review by dsgrieve — APPROVED** [review]

**Pull request merged**
